lib/_base_viewmodels.dart

```diff
@@ -66,63 +82,209 @@ abstract class ReactiveViewModel extends BaseViewModel {
+
+  Completer _futuresCompleter;
+  int _futuresCompleted;
+
+  void _initialiseData() {
+    _errorMap = Map<String, bool>();
+    _dataMap = Map<String, dynamic>();
+    _futuresCompleted = 0;
+  }
+
+  Future runFutures() {
+    _futuresCompleter = Completer();
+    _initialiseData();
+    notifyListeners();
+
+    for (var key in futuresMap.keys) {
+      runBusyFuture(futuresMap[key](), busyObject: key).then((futureData) {
+        _dataMap[key] = futureData;
+        notifyListeners();
+        onData(key);
+        _incrementAndCheckFuturesCompleted();
+      }).catchError((error) {
+        _errorMap[key] = true;
+        setBusyForObject(key, false);
+        onError(key: key, error: error);
+        notifyListeners();
+        _incrementAndCheckFuturesCompleted();
+      });
+    }
+
+    return _futuresCompleter.future;
+  }
+
+  void _incrementAndCheckFuturesCompleted() {
+    _futuresCompleted++;
+    if (_futuresCompleted == futuresMap.length &&
+        !_futuresCompleter.isCompleted) {
+      _futuresCompleter.complete();
+    }
+  }
+
+  void onError({String key, error}) {}
+
+  void onData(String key) {}
+}
+
+/// Provides functionality for a ViewModel to run and fetch data using multiple streams
+
+abstract class MultipleStreamViewModel extends _MultiDataSourceViewModel {
+  Map<String, Stream> get streamsMap;
+  Map<String, StreamData> _streamDataMap;
+  Map<String, StreamData> get streamDataMap => _streamDataMap;
+  Map<String, dynamic> get dataMap {
+    Map<String, dynamic> _result = Map<String, dynamic>();
+    _streamDataMap.forEach((key, streamData) => _result[key] = streamData.data);
+    return _result;
+  }
+
+  Map<String, dynamic> get errorMap {
+    Map<String, dynamic> _result = Map<String, dynamic>();
+    _streamDataMap
+        .forEach((key, streamData) => _result[key] = streamData.hasError);
+    return _result;
+  }
+
+  void _initialiseData() {
+    _streamDataMap = Map<String, StreamData>();
+  }
+
+  void runStreams() {
+    _initialiseData();
+    notifyListeners();
+    for (var key in streamsMap.keys) {
+      _streamDataMap[key] = setupStream(streamsMap[key]);
+      notifyListeners();
+    }
+  }
+}
+
+abstract class StreamViewModel<T> extends _SingleDataSourceViewModel<T> {
+  Stream<T> get stream;
+  // Holds the data, dataReady, hasError, isBusy properties
+  StreamData streamData;
+  get data => streamData.data;
+  get dataReady => streamData.dataReady;
+  get hasError => streamData.hasError;
+  get isBusy => streamData.isBusy;
+
+  //TODO: Integrate read/write for onData, onCancel, onSubscribed, onError and transformData properties
+  //TODO: Figure out a more elegant way of integrating StreamData
+  //TODO: Get this to work with generic stream types again
+  //TODO: Possibly have a default setbusy for overwrittenn lifecycle events
+
+  Future initialise() async {
+    streamData = setupStream(stream);
+  }
+}
+
+class StreamData<T> extends _SingleDataSourceViewModel<T> {
```

Comment:
We should make this class private so it's not available for use outside of this file